src/uu/cksum/src/cksum.rs

```diff
@@ -391,47 +394,170 @@ fn calculate_length(algo_name: &str, length: usize) -> UResult<Option<usize>> {
+
+                let (calculated_checksum, _) = digest_read(&mut algo, &mut file_reader, bits)
+                    .map_err_context(|| "failed to read input".to_string())?;
+                if expected_checksum == calculated_checksum {
+                    println!("{}: OK", filename_to_check);
+                } else {
+                    println!("{}: FAILED", filename_to_check);
+                    failed_cksum += 1;
+                }
+            } else {
+                bad_format += 1;
+            }
+        }
+    }
+
+    // not a single line correctly formatted found
+    // return an error
+    if !properly_formatted {
         return Err(io::Error::new(
-            io::ErrorKind::InvalidInput,
-            "the --binary and --text options are meaningless when verifying checksums",
+            io::ErrorKind::Other,
+            "no properly formatted checksum lines found",
         )
         .into());
     }
-    Ok((tag, asterisk))
+
+    // strict means that we should have an exit code.
+    if strict && bad_format > 0 {
+        set_exit_code(1);
+    }
+
+    // if any incorrectly formatted line, show it
+    match bad_format.cmp(&1) {
+        Ordering::Equal => {
+            show_warning_caps!("{} line is improperly formatted", bad_format);
+        }
+        Ordering::Greater => {
+            show_warning_caps!("{} lines are improperly formatted", bad_format);
+        }
+        Ordering::Less => {}
+    };
+
+    // if we have any failed checksum verification, we set an exit code
+    if failed_cksum > 0 || failed_open_file > 0 {
+        set_exit_code(1);
+    }
+
+    match failed_open_file.cmp(&1) {
+        Ordering::Equal => {
+            show_warning_caps!("{} listed file could not be read", failed_open_file);
+        }
+        Ordering::Greater => {
+            show_warning_caps!("{} listed files could not be read", failed_open_file);
+        }
+        Ordering::Less => {}
+    }
+
+    match failed_cksum.cmp(&1) {
+        Ordering::Equal => {
+            show_warning_caps!("{} computed checksum did NOT match", failed_cksum);
+        }
+        Ordering::Greater => {
+            show_warning_caps!("{} computed checksums did NOT match", failed_cksum);
+        }
+        Ordering::Less => {}
+    };
+
+    Ok(())
 }
 
 #[uucore::main]
 pub fn uumain(args: impl uucore::Args) -> UResult<()> {
     let matches = uu_app().try_get_matches_from(args)?;
 
+    let check = matches.get_flag(options::CHECK);
+
     let algo_name: &str = match matches.get_one::<String>(options::ALGORITHM) {
         Some(v) => v,
-        None => ALGORITHM_OPTIONS_CRC,
-    };
-
-    let input_length = matches.get_one::<usize>(options::LENGTH);
-
-    let length = match input_length {
-        Some(length) => calculate_length(algo_name, *length)?,
-        None => None,
+        None => {
+            if check {
+                // if we are doing a --check, we should not default to crc
+                ""
```

Comment:
Observe that `algo_name` is set here to a string that is not a valid algorithm name. (Important for the next comment.)

Reply:
This is probably the cause for this panic:

```console
$ ../gnu/src/cksum -a md5 --untagged README.md | tee foo.sums 
e5773576fc75ff0f8eba14f61587ae28  README.md
$ ../gnu/src/cksum -c foo.sums 
../gnu/src/cksum: foo.sums: no properly formatted checksum lines found
[$? = 1]
$ cargo run -q cksum -c foo.sums 
thread 'main' panicked at 'internal error: entered unreachable code: unknown algorithm: clap should have prevented this case', src/uu/cksum/src/cksum.rs:141:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[$? = 101]
```

Reply:
I think this is a similar instance of the same issue, which is still present:
```console
$ echo 'SHA4 (README.md) = 00000000' > foo.sums 
$ cksum -c foo.sums 
cksum: foo.sums: no properly formatted checksum lines found
[$? = 1]
$ cargo run -q cksum -c foo.sums 
thread 'main' panicked at src/uu/cksum/src/cksum.rs:142:14:
internal error: entered unreachable code: unknown algorithm: clap should have prevented this case
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[$? = 101]
```